version: "3.8"
services:
    grafana:
        image: grafana/grafana:10.3.3
        depends_on:
            db:
                condition: service_healthy
            loki:
                condition: service_healthy
        restart: always
        environment:
            - GF_PATHS_PROVISIONING=/etc/grafana/provisioning
            - GF_SERVER_ROOT_URL=https://grafana.example.com/
        volumes:
            - grafana_storage:/var/lib/grafana
            - ./grafana.ini:/etc/grafana/grafana.ini:ro
            - ./datasources:/etc/grafana/provisioning/datasources:ro
        networks:
            - traefik-public
            - default
        labels:
            - traefik.enable=true
            - traefik.docker.network=traefik-public
            - traefik.http.routers.grafana.entryPoints=https
            - traefik.http.routers.grafana.rule=Host(`grafana.example.com`)
            - traefik.http.routers.grafana.tls=true
            - traefik.http.routers.grafana.tls.certResolver=myResolver
            - traefik.http.services.grafana.loadBalancer.server.port=3000

    db:
        image: postgres:16.2
        restart: always
        healthcheck:
            test: pg_isready -d grafana
            interval: 5s
            timeout: 5s
            retries: 30
        environment:
            - POSTGRES_USER=grafana
            - POSTGRES_PASSWORD=${POSTGRES_GRAFANA_PASSWORD}
            - TZ=Europe/Moscow
        volumes:
            - postgres:/var/lib/postgresql/data:delegated

    loki:
        build:
            context: .
        restart: always
        healthcheck:
            test: curl -f http://127.0.0.1:3100/ready
            interval: 5s
            timeout: 5s
            retries: 30
        command: -config.file=/etc/loki/loki.yaml
        environment:
            - TZ=Europe/Moscow
        volumes:
            - loki:/data/loki
            - ./loki.yaml:/etc/loki/loki.yaml:ro

volumes:
    grafana_storage: null
    postgres: null
    loki: null

networks:
    traefik-public:
        external: true